design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,286 @@
+# Enhancing observability in Kanister
+
+<!-- toc -->
+- [Motivation](#motivation)
+- [Scope](#scope)
+- [Detailed Design](#detailed-design)
+  - [Architecture](#architecture)
+    - [Text description](#text-description)
+  - [Low Level APIs](#low-level-apis)
+    - [Metrics Package](#metrics-package)
+      - [Example Initialization Steps for a new CounterVec metric](#example-initialization-steps-for-a-new-countervec-metric)
+    - [Consumer Package](#consumer-package)
+- [Testing](#testing)
+<!-- /toc -->
+
+This document proposes the changes required within Kanister to integrate
+[Prometheus](https://github.com/prometheus/prometheus) and enhance the observability 
+of action sets. 
+
+## Motivation
+Kanister controller already has a registered metrics endpoint `/metrics`. There are no metrics exported other than the default Prometheus metrics that the default handler provides. Adding metrics to track the 
+ActionSets and Blueprints workflow will help improve the overall observability. 
+
+To achieve this, we need to build a framework for exporting metrics from the Kanister controller, and to 
+start with, export some metrics to Prometheus. 
+
+Some example metrics include:
+ActionSets succeeded
+ActionSets failed
+Action duration
+Phase duration, etc.
+
+
+## Scope
+1. Design a framework that allows us to export new
+Kanister metrics to Prometheus easily. 
+2. Add a few fundamental metrics related to ActionSets and Blueprints to start with.
+
+
+## Detailed Design
+
+### Architecture
+
+![Alt text](Metrics_design.png?raw=true "Prometheus Integration Design")
+
+#### Text description
+1. The initializer of the consumer package calls new_metrics, a helper method that talks to Kanister’s metrics package. The result is a new metrics struct that owns all the prometheus collectors.
```

Comment:
From the perspective of the most important audience, a Kanister developer, I believe we should refer to these as Prometheus **metrics**, not **collectors**.
And for consistency in this doc, we should capitalize Prometheus.

A `Collector` is an internal interface and refers more to the side of the metrics being collected by the Prometheus code that is invoked when the `/metrics` endpoint is scraped. But let's keep it simple for the Kanister developer.
(Yes, it's also the type returned for all metrics and material for unit tests, etc.,
but for getting Kanister developers to add Prometheus metrics, let's not complicate it.)

See overview terminology used at https://prometheus.io/docs/introduction/overview/

```suggestion
1. The initializer of the consumer package calls new_metrics, a helper method that talks to Kanister’s metrics package. The result is a new metrics struct that owns all the Prometheus metrics.
```

Reply:
Yes, exposing this as "Prometheus Metrics" makes sense. I've made the change throughout the document.